### PR TITLE
blocksync: verification rework- part 1

### LINKED ISSUE #4457: Light client verification for fast sync
Fast sync should use the light client for block verification.

### PR DESCRIPTION
**Pull request by jmalicevic**
Closes #4457 

The blocksync verification now resembles the verification done by the light client (without witness verification and relying on a trusted period). 

- We start from a trusted state at height H. If the LastCommit of the block at height H+1 does not verify block at H, we can dismiss the block at H+1. If it does, we use the LastCommit of the block at H+2 to verify the block at H+1. The validator set of the LastCommit at block H+2 is verified against the NextValidators stored in the trusted state (at height H).

### CODE REVIEW

File: internal/blocksync/pool.go
Comment (sergio-mena) [inline internal/blocksync/pool.go:544]: Consider renaming this to `TrustedBlockData` or something similar

File: internal/blocksync/reactor.go
Comment (sergio-mena) [inline internal/blocksync/reactor.go:552]: Do you mean `r.logger.Error(` ?

Comment (sergio-mena) [reply internal/blocksync/reactor.go:552]: Also, please add a meaningful message before the variables

File: internal/blocksync/reactor.go
Comment (sergio-mena) [inline internal/blocksync/reactor.go:559]: But this overrides the assignment in line 544...

File: internal/blocksync/reactor.go
Comment (sergio-mena) [inline internal/blocksync/reactor.go:544]: If we haven't synchronized any block yet, and we ran statesync, don't we risk loading and incomplete block here from the block store?
Well, maybe it doesn't matter because we only use it for verification, not for `ApplyBlock`

File: internal/blocksync/reactor.go
Comment (sergio-mena) [inline internal/blocksync/reactor.go:561]: But, if we came from line 559 (i.e., `else` in line 541), the blocks haven't been verified... am I missing something?

File: internal/blocksync/reactor.go
Comment (sergio-mena) [inline internal/blocksync/reactor.go:541]: I have a concern here: I think the code in this `else` should be on its own `if` before line 509. Let me know if you want to discuss.

File: internal/blocksync/verify.go
Comment (sergio-mena) [inline internal/blocksync/verify.go:85]: Someone with a _deep_ knowledge of light client verification algorithm should review this part

File: internal/blocksync/verify.go
Comment (sergio-mena) [inline internal/blocksync/verify.go:30]: Shouldn't the check be more strict? (a +1 check)

File: internal/store/store.go
Comment (sergio-mena) [inline internal/store/store.go:129]:
```suggestion
// LoadBlockProto returns the block with the given height in protobuf.
```

File: internal/store/store.go
Comment (sergio-mena) [inline internal/store/store.go:159]: Could we refactor this method to use the new one, so that we don't duplicate code?

File: internal/store/store.go
Comment (sergio-mena) [inline internal/store/store.go:285]:
```suggestion
// LoadBlockCommitProto returns the Commit for the given height in protobuf.
```

File: internal/store/store.go
Comment (sergio-mena) [inline internal/store/store.go:307]: Same comment (please call `LoadBlockCommitProto` from here to avoid code duplication)

**Review by sergio-mena — COMMENTED** [review]
Thanks for posting this for review. I understand there are some small part to be finished. But I am happy I can now see what we will need to "merge with" from vote extensions perspective.
Please ping me if you have questions on any of my comments.

File: internal/blocksync/verify.go
Comment (marbar3778) [inline internal/blocksync/verify.go:12]: can the light client verifier be reused here instead of duplicating?

File: internal/blocksync/reactor.go
Comment (jmalicevic) [reply internal/blocksync/reactor.go:510]: Actually I am not 100% sure that the way it was done before was correct. 
In the previous scenario, `state` is the state existing at height `H `and then we want to verify the `first` block at height` H + 1`. So one should verify the `first` block with the validator set for the height of that block. But the `validator` field of the state relates to the validators that validate that particular state, not necessarily the set that will be used to validate the state at the next height.

So here I take the validator set of` H + 1` to verify `newBlock` , and at the same time [here](https://github.com/tendermint/tendermint/blob/4b83be646c140fb50391acc40384dcb1cd2e10e6/internal/blocksync/verify.go#L95): I use it to verify that these validators indeed signed the commit of `verifyBlock.lastCommit`(on height H + 2). It is assumed that we trust the state but we need the last commit of verifyBlock to verify newBlock. This was the only way I could think of to at least somehow increase the trust in verifyBlock.

File: internal/blocksync/reactor.go
Comment (jmalicevic) [reply internal/blocksync/reactor.go:552]: Changed this to a valid error and added a request redo here.

File: internal/blocksync/reactor.go
Comment (jmalicevic) [reply internal/blocksync/reactor.go:561]: Good catch, result of copy pasting..basically we need to contiunue the loop whenever verification fails and keep waiting for a new block!

Comment (jmalicevic) [reply internal/blocksync/reactor.go:561]: This is why adding a test with a scenario that should not validate is important. It would have caught all these cases.

File: internal/blocksync/verify.go
Comment (jmalicevic) [reply internal/blocksync/verify.go:12]: We cannot simply call the light client `VerifyAdjacent` here as the light client function has time constraints. It verifies whether a block is within the trusted period which may not be true if we are blocksync-ing. That being said, we can always alter the `VerifyAdjacent` function in the light client to conditionally check whether we are are calling it from blocksync or the light client. For now, I added it separately until we converge to the desired verification function.

File: internal/blocksync/reactor.go
Comment (sergio-mena) [reply internal/blocksync/reactor.go:507]: OK, so if you followed your explanation correctly, the old code, instead of `state.Validators.VerifyCommitLight(...)` should have been doing `state.NextValidators.VerifyCommitLight(...)` all along. Is that so?

File: internal/blocksync/reactor.go
Comment (cason) [inline internal/blocksync/reactor.go:196]: It seems you can remove this part of the code, since `block` is now a proto version.

File: internal/blocksync/reactor.go
Comment (cason) [inline internal/blocksync/reactor.go:243]: Is the peer information not useful?

**Review by cason — APPROVED** [review]

File: internal/blocksync/reactor.go
Comment (jmalicevic) [reply internal/blocksync/reactor.go:190]: In a previous version, where I was reading the commit in addition to the block (what is now done when reading the extending commit), when I had both being returned as Objects and then converted to Proto again, the tests were constantly timing out. In the meantime I removed the commit but somehow it made sense to me to only read Proto instead of converting it back and forth so I left this change. This logic should probably be applied for the extended commit now that it is also read. Especially if the extended commit ends up being larger, we might have a problem again.

File: internal/blocksync/reactor.go
Comment (jmalicevic) [reply internal/blocksync/reactor.go:580]: Is the state after ApplyBlock as well at the height of a subsequent NewBlock ? Apply block sets the `state.validators` to the `previousState.nextValidators`.

File: internal/blocksync/pool.go
Comment (jmalicevic) [reply internal/blocksync/pool.go:241]: Yes, I think we should skip the first case. The start height is always at least 1 (it is the height we are currently syncing with, so it should be included in the count) so technically the %100 will not be 0 for 100 blocks, rather for 101 and this caused some tests to be flaky. But I will add a check to skip the first sync. @sergio-mena , @cmwaters  does that sound ok?

File: internal/blocksync/reactor.go
Comment (cmwaters) [reply internal/blocksync/reactor.go:554]: After `ApplyBlock`, `state.LastBlockHeight` is `NewBlock.Height` and state is thus for `NewBlock.Height + 1`

File: internal/blocksync/reactor.go
Comment (jmalicevic) [inline internal/blocksync/reactor.go:80]: @sergio-mena , I reverted this temporarily to verify the behaviour of the rest of the code. Then we can re-implement your change and use an interface.

File: internal/blocksync/reactor.go
Comment (jmalicevic) [reply internal/blocksync/reactor.go:604]: I think as long as we have the data needed for verification this is fine. But, I discovered a potential bug here. If we indeed start from scratch and we have no previous commit stored (initial block), we will have a Null pointer exception on the commit. Working on fixing this, this case was not properly handled.

File: internal/blocksync/pool.go
Comment (jmalicevic) [reply internal/blocksync/pool.go:561]: The name was left over from the rpevious implementation. Renamed it.

File: internal/blocksync/reactor.go
Comment (jmalicevic) [reply internal/blocksync/reactor.go]: I confirmed this in the doc, `ApplyBlock` returns state for H + 1 and changed this  to use` state.Validators`.

File: internal/blocksync/pool.go
Comment (cmwaters) [reply internal/blocksync/pool.go]: I think this is fine

File: internal/blocksync/reactor.go
Comment (cmwaters) [reply internal/blocksync/reactor.go:190]: > the tests were constantly timing out

That seems strange. And you were able to track it to changing the struct.

File: internal/blocksync/pool.go
Comment (sergio-mena) [reply internal/blocksync/pool.go]: Sounds good

File: internal/blocksync/reactor.go
Comment (sergio-mena) [reply internal/blocksync/reactor.go:80]: I'm OK both ways. The change to using the interface was more of an exercise to make sure I had understood how the block storage is used.

File: internal/blocksync/pool.go
Comment (sergio-mena) [inline internal/blocksync/pool.go:241]: I may be missing the point here but, like it is now (and for a `startHeight` of 1), won't the condition become true for height 2?

File: internal/blocksync/reactor.go
Comment (sergio-mena) [inline internal/blocksync/reactor.go:551]: But this can never happen, can it?
Did you rather mean `r.lastTrustedBlock.block`?

File: spec/abci++/abci++_methods_002_draft.md
Comment (sergio-mena) [inline spec/abci++/abci++_methods_002_draft.md:292]: Why this change?

File: internal/blocksync/reactor.go
Comment (jmalicevic) [reply internal/blocksync/reactor.go:550]: I am fairly confident in this but I changed this to be `state.LastBlockHeight !=0` as it is more in line with the rest of the code (if you look at the top, `state` is being initialized with `r.initialState` and `r.initialState` is supposed to be immutable after the first time it is set.

File: internal/blocksync/reactor.go
Comment (jmalicevic) [reply internal/blocksync/reactor.go:551]: Yes, you are right.

File: internal/blocksync/reactor.go
Comment (jmalicevic) [reply internal/blocksync/reactor.go:569]: Nice catch, this is a bad copy/paste. This is state.LastBlockHeight

File: internal/blocksync/pool.go
Comment (jmalicevic) [reply internal/blocksync/pool.go:241]: I f we want to really count every 100 blocks and we include the block at `initialHeight` (which is also synced), the correct condition is in fact `(pool.height - pool.startHeight + 1) % 100`. Additionally, for tests not to time out, we need to set the height of the chain ( `pool.maxPeerHeigth`) to be 100 + the maximum initial height of the testnet nodes. Otherwise the sync rate is not calculated properly.

File: internal/blocksync/reactor.go
Comment (sergio-mena) [reply internal/blocksync/reactor.go]: I was rather referring to the fact that the seen commit is queried without a height as it is always the most recent seen.

File: spec/abci++/abci++_methods_002_draft.md
Comment (sergio-mena) [reply spec/abci++/abci++_methods_002_draft.md:292]: Could be, the last three ring a bell TBH. But since I have several PRs advancing in parallel, I lost track what's in and what not. Sorry if this caused you trouble

File: internal/blocksync/reactor.go
Comment (jmalicevic) [reply internal/blocksync/reactor.go]: Ah I see.. well, if the last seen commit is not equal to state.lastBlockHeight then we have a problem. Maybe we should check this first in any case.

File: internal/blocksync/reactor.go
Comment (jmalicevic) [reply internal/blocksync/reactor.go:191]: Yes, I left it like this. I think in any case it is more elegant to not convert it back and forth when not needed.

File: internal/blocksync/reactor.go
Comment (jmalicevic) [reply internal/blocksync/reactor.go]: Just committed the check, should be fine now.

**Comment by cmwaters** [comment]
This work, I believe, is a little on hold at the moment as the team focus on v0.35 and v0.36 releases. Hopefully we can return to this soon and get it in for v0.37

**Pull request closed without merging**
